### PR TITLE
new: CNS rql suggest and search

### PR DESCRIPTION
**Pull request by jping0929**

### CODE REVIEW

**Comment by jping0929** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "1164c8b84235d5732622f1296875cc55836baf54",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "38741268723a0d6d2b12e0046b3f85ee9e757985",
    "pipeline": "master"
  }
]
```

File: custom_validations.go [hunk 1]
```diff
@@ -1145,3 +1145,19 @@ func ValidateCachedFlowReport(cachedFlowReport *CachedFlowReport) error {
 
 	return nil
 }
+
+// ValidateDNSLookupReport validates a DNSLookupReport.
+func ValidateDNSLookupReport(report *DNSLookupReport) error {
+
+	switch {
+	case report.Namespace == "" && report.ProcessingUnitNamespace != "":
```
Comment (t00f) [inline custom_validations.go:1153]: Add a comment here to explain that we temporarily assign the namespace from the PUNamespace until the enforcer is reporting the correct namespace.

File: custom_validations.go [hunk 2]
```diff
@@ -1145,3 +1145,19 @@ func ValidateCachedFlowReport(cachedFlowReport *CachedFlowReport) error {
...
+		report.Namespace = report.ProcessingUnitNamespace
+		report.ProcessingUnitNamespace = ""
+	case report.Namespace != "" && report.ProcessingUnitNamespace != "":
+		return makeValidationError("namespace", "Both 'namespace' and 'processingUnitNamespace' cannot be set")
+	case report.Namespace == "" && report.ProcessingUnitNamespace == "":
+		return makeValidationError("namespace", "Attribute 'namespace' is required")
+	}
```
Comment (t00f) [inline custom_validations.go:1160]: Looks good :) I would have used 3 if conditions instead of an empty switch though

**Comment by t00f** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "4ae970b50529f72f7fec034310ec9430e8c08e7c",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "8131e154cfaf8f769d2c17c121f408f1001e23d0",
    "pipeline": "master"
  }
]
```

**Review by t00f — DISMISSED** [review]

**Comment by jping0929** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "befa869e278aaeafe3fb0a1932b111fdbde00b83",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "8131e154cfaf8f769d2c17c121f408f1001e23d0",
    "pipeline": "master"
  }
]
```

File: specs/cnssearch.spec [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    description: Indicates if the search has been saved.
+    type: boolean
+    exposed: true
+    omit_empty: true
+
+  - name: searchType
+    description: Type of search request. Should set to be network.
+    type: string
+    exposed: true
+    omit_empty: true
```
Comment (ericrpowers) [inline specs/cnssearch.spec:77]: We can probably define the types allowed here unless they are not well-defined. Something like:
```
  - name: searchType
    description: Type of search request.
    type: enum
    exposed: true
    allowed_choices:
    - network
    - other-type
    default_value: network
```

File: specs/cnssuggestion.spec [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    subtype: string
+    example_value:
+    - id
+    - action
+
+  - name: translate
+    description: Required by Prisma Cloud. Always set to false.
+    type: boolean
+    exposed: true
+    default_value: false
```
Comment (ericrpowers) [inline specs/cnssuggestion.spec:46]: We can probably just avoid exposing this if there is no choice in its setting.

File: specs/cnssuggestion.spec [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  description: Provides query suggestions for Prisma Cloud's investigate page.
+
+# Attributes
+attributes:
+  v1:
+  - name: needsOffsetUpdate
+    description: Required by Prisma Cloud. Always set to true.
+    type: boolean
+    exposed: true
+    default_value: true
```
Comment (ericrpowers) [inline specs/cnssuggestion.spec:17]: We can probably just avoid exposing this if there is no choice in its setting.

File: specs/cnssearch.spec [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  resource_name: cnssearches
+  entity_name: CNSSearch
+  package: karl
+  group: core/rql
+  description: Provide search results for Prisma Cloud's investigate page.
+
+# Attributes
+attributes:
+  v1:
+  - name: ID
```
[the remark this answers is not in the document]
Comment (jping0929) [reply specs/cnssearch.spec:13]: 👍 their ID is lower cased so should probably leave it as it is?

File: specs/cnssuggestion.spec (hunk 5 above)
Comment (jping0929) [reply specs/cnssuggestion.spec:17]: if we do not expose them, wouldn't the returned JSON omit these values?

Comment (ericrpowers) [reply specs/cnssuggestion.spec:17]: Ahhh, this is what is given back to Prisma Cloud, not us. Then this is fine as with the other above.

File: specs/cnssearch.spec (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericrpowers) [reply specs/cnssearch.spec:13]: Ahh, then ID can be left alone. What about `description` and/or `name`? I did not really grasp why `name` needed to be empty?

[the remark this answers is not in the document]
Comment (jping0929) [reply specs/cnssearch.spec:13]: it's something vaguely specified in the PC spec :( These attributes are not really aporeto attributes

File: specs/cnssearch.spec (hunk 3 above)
Comment (jping0929) [reply specs/cnssearch.spec:77]: For now we only need to use `network`, but the first letter is lower case... so I probably cannot use `allowed_choices` :(

**Comment by jping0929** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "abc4043ff4b5aa8f84a579392efde1f353379ce4",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "8b285a01042c8e64fe3b3fc53077903f9ef3e864",
    "pipeline": "master"
  }
]
```

**Comment by jping0929** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "8f331db601fb7c9022fb541239dfe0cbe4a62a01",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "8b285a01042c8e64fe3b3fc53077903f9ef3e864",
    "pipeline": "master"
  }
]
```

**Comment by jping0929** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "656fe5ba63a6d4852f90ce3b18b005aad554473c",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "8b285a01042c8e64fe3b3fc53077903f9ef3e864",
    "pipeline": "master"
  }
]
```

**Comment by jping0929** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "6c809026e59de4030f2e42a4b189f0698490dd15",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "8b285a01042c8e64fe3b3fc53077903f9ef3e864",
    "pipeline": "master"
  }
]
```

**Comment by jping0929** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "037f703850e4ee30b55538cfea7b57f2d84b465d",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "8b285a01042c8e64fe3b3fc53077903f9ef3e864",
    "pipeline": "master"
  }
]
```

**Comment by jping0929** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "cns-search-suggest",
    "component": "backend",
    "pr-id": "962",
    "commit-sha": "736e488450e5cd55f384f59869059b345e7437b2",
    "pipeline": "master"
  },
  {
    "project": "cns-search-suggest",
    "component": "gaia",
    "pr-id": "592",
    "commit-sha": "8b285a01042c8e64fe3b3fc53077903f9ef3e864",
    "pipeline": "master"
  }
]
```

**Review by t00f — APPROVED** [review]

**Pull request merged**